src/Stack/Fetch.hs

```diff
@@ -554,6 +533,69 @@ fetchPackages' mdistDir toFetchAll = do
                 S.writeFile cabalFP $ tfCabal toFetch
 
                 atomically $ modifyTVar outputVar $ Map.insert ident destDir
+            $logWarn $ mconcat $ map (\(path, entryType) -> "Unexpected entry type " <> entryType <> " for entry " <> T.pack path) unexpectedEntries
+
+-- | Internal function used to unpack tarball.
+--
+-- Takes a path to a .tar.gz file, the name of the directory it should contain,
+-- and a destination folder to extract the tarball into. Returns unexpected
+-- entries, as pairs of paths and descriptions.
+untar :: FilePath -> FilePath -> FilePath -> IO [(FilePath, T.Text)]
```

Comment:
I would prefer to see `Path` being used here as much as possible / reasonable. IMO the distinction between the different `Path` types makes code (especially type signatures) more readable and slightly more robust.

You can find utils for working with `Path` in `src/Path` (especially the amazing `Path.Extra.toFilePathNoTrailingSep` ;)) and in the `path-io` package.